drgn_tools/meminfo.py

```diff
@@ -0,0 +1,755 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helpers for dumping memory usage information and statistics
+"""
+import argparse
+from typing import Dict
+from typing import Iterator
+from typing import List
+
+from drgn import Object
+from drgn import Program
+from drgn.helpers.linux import list_for_each_entry
+from drgn.helpers.linux.percpu import percpu_counter_sum
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.mm import totalram_pages
+from drgn_tools.util import enum_value_get
+from drgn_tools.util import get_uts
+from drgn_tools.util import has_member
+
+
+__all__ = ("show_all_meminfo", "get_all_meminfo")
+
+
+# Machine info
+_IS_X86_64 = False
+_IS_AARCH_64 = False
+_UEK_VERSION = 5
+
+# Basic memory subsystem parameters are in the following:
+# The number of NUMA
+_NODES_SHIFT = 10
+_MAX_NUM_NODES = 1 << _NODES_SHIFT
+_MAX_NUM_ZONES = 0
+
+# Page settings.
+_PAGE_SIZE = 0
+_PAGE_SHIFT = 0
+_PMD_SHIFT = 0
+_HPAGE_SHIFT = 0
+
+_HPAGE_PMD_SHIFT = 0
+_HPAGE_PMD_ORDER = 0
+_HPAGE_PMD_NR = 0
+
+# vmalloc memory range.
+_VMALLOC_START = 0
+_VMALLOC_END = 0
+
+# Using five bits for the type means that the maximum number of
+# swapcache pages is 27 bits * on 32-bit-pgoff_t architectures.
+_MAX_SWAPFILES_SHIFT = 5
+# NUMA node memory migration support
+_SWP_DEVICE_NUM = 2
+# NUMA node memory migration support
+_SWP_MIGRATION_NUM = 2
+# Handling of hardware poisoned pages with memory corruption.
+_SWP_HWPOISON_NUM = 1
+
+# |MAX_SWAPFILES| defines the maximum number of swaptypes.
+_MAX_SWAPFILES = (
+    (1 << _MAX_SWAPFILES_SHIFT)
+    - _SWP_DEVICE_NUM
+    - _SWP_MIGRATION_NUM
+    - _SWP_HWPOISON_NUM
+)
+
+# The index for unknown stats ID for node stats and zone stats.
+UNKNOWN_GLOBAL_STAT_ID = -1
+
+
+def print_val_kb(text: str, num: int) -> None:
+    """
+    Produce the formatted output that matches the output in /proc/meminfo.
+
+    param text: The output item's name, e.g, "MemTotal"
+    param num: The output item's value
+    """
+    if num < 0:
+        return
+    text += ":"
+    num_kb = num << (_PAGE_SHIFT - 10)
+    print("%s%s kB" % (text.ljust(16, " "), str(num_kb).rjust(8, " ")))
+
+
+def parse_machine_info(prog: Program) -> None:
```

Comment:
I'd suggest two approaches for this function - use whichever you like better.

First option is to have it return a dictionary mapping the constant name to its value. You could name the function `get_mm_constants()` or something like that. This can avoid storing everything in a global variable, and you can even use the prog.cache to cache it.

Second option is to simply compute the constants where you need them. Aside from the difficult vmalloc constants, I think most of them are easy to just compute whenever you want them.

Reply:
I adopted the first option because these variables are used in many functions. Caching them can save some memory lookups.